core/src/main/scala/offheap/Region.scala

```diff
@@ -29,10 +29,18 @@ trait Region extends Allocator {
     checkOpen
     if (Checked.MEMORY) Sanitizer.unregister(id)
   }
-  def reallocate(addr: Addr, size: Size): Addr =
-    throw new UnsupportedOperationException
+  def reallocate(oldAddr: Addr, oldSize: Size, newSize: Size): Addr = {
+    checkOpen
+    if (newSize <= oldSize)
+      oldAddr
```

Comment:
I think most implementations of realloc in the C library would either leave the address the same but actually free the remaining / discarded space, or perform a new allocation, copy and complete free of the original memory. Would you be open to the first approach? 

And similarly for the 'else' block below, do you need to explicitly free the old memory so you don't get too much fragmentation in your region? I might very well be missing the desired intent of the commit though.